### PR TITLE
Check backend disk has available capactiy by storage medium before create table

### PR DESCRIPTION
**Pull request by WingsGo**
Currently we choose BE random without check disk is available, the create table will failed until create tablet task is sent to BE and BE will check is there has avaliable capcity to create tablet.So Check
backend disk available by storage medium will reduce unnecessary RPC call.

### CODE REVIEW

**Comment by WingsGo** [comment]
For #3402

File: fe/src/main/java/org/apache/doris/catalog/Catalog.java [hunk 1]
```diff
@@ -4894,8 +4894,7 @@ public void renameTable(Database db, OlapTable table, TableRenameClause tableRen
 
         // check if rollup has same name
         if (table.getType() == TableType.OLAP) {
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/catalog/Catalog.java:4896]: We can just remove the `if (table.getType() == TableType.OLAP)`. The argument is already `OlapTable`

File: fe/src/main/java/org/apache/doris/catalog/Catalog.java [hunk 2]
```diff
@@ -4894,8 +4894,7 @@ public void renameTable(Database db, OlapTable table, TableRenameClause tableRen
...
-            OlapTable olapTable = (OlapTable) table;
-            for (String idxName: olapTable.getIndexNameToId().keySet()) {
+            for (String idxName: table.getIndexNameToId().keySet()) {
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/catalog/Catalog.java:4897]:
```suggestion
            for (String idxName : table.getIndexNameToId().keySet()) {
```

File: fe/src/main/java/org/apache/doris/system/Backend.java [hunk 3]
```diff
@@ -346,6 +347,23 @@ public double getMaxDiskUsedPct() {
         return maxPct;
     }
 
+    public boolean checkDiskExceedLimitByStorageMedium(TStorageMedium storageMedium) {
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/system/Backend.java:350]: Add comment to explain the return value of this method.
the name `checkDiskExceedLimitByStorageMedium` is a little bit confused.

File: fe/src/main/java/org/apache/doris/system/Backend.java [hunk 4]
```diff
@@ -346,6 +347,23 @@ public double getMaxDiskUsedPct() {
...
+        // no available disk with storage medium
+        if (getDiskNumByStorageMedium(storageMedium) <= 0) {
+            return true;
+        }
+        ImmutableMap<String, DiskInfo> disks = disksRef.get();
+        boolean exceedLimit = false;
+        for (DiskInfo diskInfo : disks.values()) {
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/system/Backend.java:357]: Is this logic correct? If all disks are offline or is not with specified storage medium, then this method will return false?

File: fe/src/main/java/org/apache/doris/system/SystemInfoService.java [hunk 5]
```diff
@@ -443,7 +390,7 @@ public void releaseBackends(String clusterName, boolean isReplay) {
         }
 
         List<List<Backend>> hostList = Lists.newArrayList(hostBackendsMapInCluster.values());
-        Collections.sort(hostList, hostBackendsListComparator);
+        hostList.sort(hostBackendsListComparator);
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/system/SystemInfoService.java:393]: Actually, the method `calculateDecommissionBackends()` is deprecated.
So could you please add `@Derepcated` to this method?

File: fe/src/main/java/org/apache/doris/system/SystemInfoService.java [hunk 6]
```diff
@@ -729,8 +639,18 @@ public void releaseBackends(String clusterName, boolean isReplay) {
...
     // use synchronized to run serially
     public synchronized List<Long> seqChooseBackendIds(int backendNum, boolean needAlive, boolean isCreate,
-            String clusterName) {
-        long lastBackendId = -1L;
+                                                       String clusterName) {
+        return seqChooseBackendIds(backendNum, needAlive, isCreate, clusterName, getClusterBackends(clusterName));
+    }
+    public synchronized List<Long> seqChooseBackendIdsByStorageMedium(int backendNum, boolean needAlive, boolean isCreate,
+                                                                      String clusterName, TStorageMedium storageMedium) {
+        final List<Backend> backends = getClusterBackends(clusterName).stream().filter(v -> !v.checkDiskExceedLimitByStorageMedium(storageMedium)).collect(Collectors.toList());
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/system/SystemInfoService.java:647]: This is not user friendly. For example, if all backends not meet the `checkDiskExceedLimitByStorageMedium()`, the user will only get the msg: "Not enough backends", but he will not know why there is no enough backend.
So could you please find a way to return the detail msg to user about the reason of failure?

**Review by morningman — CHANGES_REQUESTED** [review]
By the way, the code refactor can be done in another PR. Or the change list is too long that will be hard for the reviewer to review the core logic of your change.

And could you add some docs for this new logic? Can be in `CREATE TABLE.md`

File: fe/src/main/java/org/apache/doris/system/SystemInfoService.java (hunk 6 above)
Comment (WingsGo) [reply fe/src/main/java/org/apache/doris/system/SystemInfoService.java:647]: actually, I change the detail msg to `throw new DdlException("Failed to find enough host in all backends with storage medium is " + storageMedium + ". need: " + replicationNum`

File: fe/src/main/java/org/apache/doris/system/SystemInfoService.java (hunk 5 above)
Comment (WingsGo) [reply fe/src/main/java/org/apache/doris/system/SystemInfoService.java:393]: done

File: fe/src/main/java/org/apache/doris/system/Backend.java (hunk 4 above)
Comment (WingsGo) [reply fe/src/main/java/org/apache/doris/system/Backend.java:357]: yes, you are right, I fixed it, and should we change the logic in be, just remove the judgement `_available_storage_medium_type_count == 1`?

```
for (auto& it : _store_map) {
            if (it.second->is_used()) {
                if (_available_storage_medium_type_count == 1
                    || it.second->storage_medium() == storage_medium) {
                    stores.push_back(it.second);
                }
            }
        }
```

File: fe/src/main/java/org/apache/doris/system/Backend.java (hunk 3 above)
Comment (WingsGo) [reply fe/src/main/java/org/apache/doris/system/Backend.java:350]: done

**Comment by WingsGo** [comment]
> By the way, the code refactor can be done in another PR. Or the change list is too long that will be hard for the reviewer to review the core logic of your change.
> 
> And could you add some docs for this new logic? Can be in `CREATE TABLE.md`

Ok, I will remove the code refactor next time, it's because Catalog.java contains the core logic and I didn't remove them.By the way, doc is updated.

**Comment by WingsGo** [comment]
Hi, @morningman I have remove the code refactor to other PR, this PR is ready for review~

**Review by morningman — COMMENTED** [review]
I think we need to add a FE config like `enable_strict_storage_medium_check`, and default is false, and
can also be modified at runtime.
Because I am afraid if I upgrade to this version, it will block some existing workload which is trying to create SSD table without SSD storage medium.

After adding the config, don't forget to update the config document too.

**Comment by WingsGo** [comment]
> I think we need to add a FE config like `enable_strict_storage_medium_check`, and default is false, and
> can also be modified at runtime.
> Because I am afraid if I upgrade to this version, it will block some existing workload which is trying to create SSD table without SSD storage medium.
> 
> After adding the config, don't forget to update the config document too.

OK, done, and In order to keep the origin behavior, I remove the check in BE, just check it in FE.

File: fe/src/main/java/org/apache/doris/common/Config.java [hunk 7]
```diff
@@ -1082,5 +1082,11 @@
      */
     @ConfField(mutable = true, masterOnly = true)
     public static boolean drop_backend_after_decommission = true;
+
+    /**
+     *
+     */
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/common/Config.java:1088]: missing coment?

Comment (WingsGo) [reply fe/src/main/java/org/apache/doris/common/Config.java:1088]: done

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**